src/python/pants/backend/go/util_rules/analyze_test_sources.go

```diff
@@ -32,23 +37,83 @@ import (
 //    test framework with a list of tests, benchmarks, examples, and fuzz
 //    targets read from source files.
 //
+// https://github.com/bazelbuild/rules_go/blob/master/go/tools/builders/generate_test_main.go
+//
 
-type TestCase struct {
+type TestFunc struct {
 	Package string `json:"package"`
 	Name    string `json:"name"`
 }
 
+type Example struct {
+	Package   string `json:"package"`
+	Name      string `json:"name"`
+	Output    string `json:"output"`
+	Unordered bool   `json:"unordered"`
+}
+
 // TestSourcesMetadata contains metadata about tests/benchmarks extracted from the parsed sources.
 // TODO: "Examples" and "fuzz targets" (Go 1.18+).
 type TestSourcesMetadata struct {
 	// Names of all functions in the test sources that heuristically look like test functions.
-	Tests []*TestCase `json:"tests,omit_empty"`
+	Tests []*TestFunc `json:"tests,omitempty"`
 
 	// Names of all functions in the test sources that heuristically look like benchmark functions.
-	Benchmarks []*TestCase `json:"benchmarks,omit_empty"`
+	Benchmarks []*TestFunc `json:"benchmarks,omitempty"`
+
+	// Testable examples. Extracted using "go/doc" package.
+	Examples []*Example `json:"examples,omitempty"`
 
 	// True if the sources already contain a `TestMain` function (which is the entry point for test binaries).
-	HasTestMain bool `json:"has_test_main"`
+	TestMain *TestFunc `json:"test_main,omitempty"`
+}
+
+// isTestFunc tells whether fn has the type of a testing function. arg
+// specifies the parameter type we look for: B, M or T.
+func isTestFunc(fn *ast.FuncDecl, arg string) bool {
+	if fn.Type.Results != nil && len(fn.Type.Results.List) > 0 ||
+		fn.Type.Params.List == nil ||
+		len(fn.Type.Params.List) != 1 ||
+		len(fn.Type.Params.List[0].Names) > 1 {
```

Comment:
It'd help to use parentheses here. It's subtle to figure out the order of operations.

Reply:
This is how the upstream wrote it. What's your stance on us deviating from the upstream?

Reply:
I think it's worth it here. More important that we can understand the fork than being identical to it.

Reply:
Added parens to the condition on the first-line so all of the `||` clauses are at "same level."